src/utils/columnUtils.ts

```diff
@@ -34,16 +42,46 @@ export function getColumnMetrics<R, SR>(metrics: Metrics<R, SR>): ColumnMetrics<
       allocatedWidths += width;
     }
 
-    const column = { ...metricsColumn, width };
+    let column: IntermediateColumn = { ...metricsColumn, width };
 
-    if (column.frozen) {
+    if (groupBy.includes(column.key)) {
+      lastFrozenColumnIndex++;
+      const level = groupBy.indexOf(column.key);
+      column = {
+        ...column,
+        frozen: true,
+        rowGroup: true,
+        groupFormatter: column.groupFormatter ?? ToggleGroupFormatter,
+        formatterOptions: {
+          groupFocusable(row: GroupRow<R>) {
+            return row.level === level;
+          }
+        }
+      };
+      columns.push(column);
+    } else if (column.frozen) {
       lastFrozenColumnIndex++;
       columns.splice(lastFrozenColumnIndex, 0, column);
     } else {
       columns.push(column);
     }
   }
 
+  // Sort groupBy columns
+  if (groupBy.length >= 0) {
```

Comment:
We can do better than this.
I think implementing sorting with `.sort()` is more flexible, we could add more sorting features in it in the future.
`groupBy` is also not `O(n^2)` anymore, which should please @qili26 I hope.
```patch
diff --git a/src/utils/columnUtils.ts b/src/utils/columnUtils.ts
index a96fde03..c7a4c8c5 100644
--- a/src/utils/columnUtils.ts
+++ b/src/utils/columnUtils.ts
@@ -1,4 +1,4 @@
-import { Column, CalculatedColumn, FormatterProps, Omit, GroupRow } from '../types';
+import { Column, CalculatedColumn, FormatterProps, GroupRow } from '../types';
 import { ToggleGroupedFormatter } from '../formatters';
 import { SELECT_COLUMN_KEY } from '../Columns';
 
@@ -26,13 +26,10 @@ export function getColumnMetrics<R, SR>(metrics: Metrics<R, SR>): ColumnMetrics<
   let allocatedWidths = 0;
   let unassignedColumnsCount = 0;
   let lastFrozenColumnIndex = -1;
-  type IntermediateColumn = Omit<Column<R, SR>, 'width'> & { width: number | undefined; rowGroup?: boolean };
-  let columns: IntermediateColumn[] = [];
+  type IntermediateColumn = Column<R, SR> & { width: number | undefined; rowGroup?: boolean };
+  const { rawGroupBy } = metrics;
 
-  // Find valid groupBy columns
-  const groupBy: readonly string[] = metrics.rawGroupBy?.filter(key => metrics.columns.find(c => c.key === key) !== undefined) ?? [];
-
-  for (const metricsColumn of metrics.columns) {
+  const columns = metrics.columns.map(metricsColumn => {
     let width = getSpecifiedWidth(metricsColumn, metrics.columnWidths, metrics.viewportWidth);
 
     if (width === undefined) {
@@ -42,45 +39,56 @@ export function getColumnMetrics<R, SR>(metrics: Metrics<R, SR>): ColumnMetrics<
       allocatedWidths += width;
     }
 
-    let column: IntermediateColumn = { ...metricsColumn, width };
+    const column: IntermediateColumn = { ...metricsColumn, width };
 
-    if (groupBy.includes(column.key)) {
-      lastFrozenColumnIndex++;
-      const level = groupBy.indexOf(column.key);
-      column = {
-        ...column,
-        frozen: true,
-        rowGroup: true,
-        groupFormatter: column.groupFormatter ?? ToggleGroupedFormatter,
-        formatterOptions: {
-          groupFocusable(row: GroupRow<R>) {
-            return row.level === level;
-          }
+    if (rawGroupBy?.includes(column.key)) {
+      const level = rawGroupBy.indexOf(column.key);
+      column.frozen = true;
+      column.rowGroup = true;
+      column.groupFormatter = column.groupFormatter ?? ToggleGroupedFormatter;
+      column.formatterOptions = {
+        groupFocusable(row: GroupRow<R>) {
+          return row.level === level;
         }
       };
-      columns.push(column);
-    } else if (column.frozen) {
+    }
+
+    if (column.frozen) {
       lastFrozenColumnIndex++;
-      columns.splice(lastFrozenColumnIndex, 0, column);
-    } else {
-      columns.push(column);
     }
-  }
 
-  // Sort groupBy columns
-  if (groupBy.length >= 0) {
-    const selectColumn = columns.find(c => c.key === SELECT_COLUMN_KEY);
-    const groupByColumns = groupBy.map(key => columns.find(c => c.key === key)!);
-    const remaningColumns = columns.filter(c => !groupByColumns.includes(c) && c !== selectColumn);
-    columns = [
-      ...groupByColumns,
-      ...remaningColumns
-    ];
+    return column;
+  });
 
-    if (selectColumn) {
-      columns.unshift(selectColumn);
+  columns.sort(({ key: aKey, frozen: frozenA }, { key: bKey, frozen: frozenB }) => {
+    // Sort select column first:
+    if (aKey === SELECT_COLUMN_KEY) return -1;
+    if (bKey === SELECT_COLUMN_KEY) return 1;
+    // Sort grouped columns second, following the groupBy order:
+    if (rawGroupBy) {
+      if (rawGroupBy.includes(aKey)) {
+        if (rawGroupBy.includes(bKey)) {
+          return rawGroupBy.indexOf(aKey) - rawGroupBy.indexOf(bKey);
+        }
+        return -1;
+      }
+      if (rawGroupBy.includes(bKey)) {
+        return 1;
+      }
     }
-  }
+    // Sort frozen columns third:
+    if (frozenA) {
+      if (frozenB) {
+        return 0;
+      }
+      return -1;
+    }
+    if (frozenB) {
+      return 1;
+    }
+    // Sort other columns last:
+    return 0;
+  });
 
   const unallocatedWidth = metrics.viewportWidth - allocatedWidths;
   const unallocatedColumnWidth = Math.max(
@@ -88,6 +96,9 @@ export function getColumnMetrics<R, SR>(metrics: Metrics<R, SR>): ColumnMetrics<
     metrics.minColumnWidth
   );
 
+  // Valid groupBy columns
+  const groupBy: string[] = [];
+
   const calculatedColumns: CalculatedColumn<R, SR>[] = columns.map((column, idx) => {
     // Every column should have a valid width as this stage
     const width = column.width ?? clampColumnWidth(unallocatedColumnWidth, column, metrics.minColumnWidth);
@@ -102,6 +113,9 @@ export function getColumnMetrics<R, SR>(metrics: Metrics<R, SR>): ColumnMetrics<
     };
     totalWidth += width;
     left += width;
+    if (column.rowGroup) {
+      groupBy.push(column.key);
+    }
     return newColumn;
   });
```
Where's my raise?